File: src/leetcode/java/two_pointers/ThreeSumClosest_16.java
package leetcode.java.two_pointers;

import java.util.Arrays;

public class ThreeSumClosest_16 {
  public int threeSumClosest(int[] nums, int target) {
    int closest = nums[0] + nums[1] + nums[2];
    if (nums.length == 3) return closest;
    Arrays.sort(nums);
    int head, tail, sum;
    for (int i = 0; i < nums.length - 2; i++) {
      if (i > 0 && nums[i] == nums[i - 1]) continue;
      head = i + 1;
      tail = nums.length - 1;
      while (head < tail) {
        sum = nums[i] + nums[head] + nums[tail];
        if (sum < target) while (head < tail && nums[++head] == nums[head - 1]);
        else if (sum > target) while (head < tail && nums[--tail] == nums[tail + 1]);
        else return sum;
        if (Math.abs(sum - target) < Math.abs(closest - target)) closest = sum;
      }
    }
    return closest;
  }
}

File: src/leetcode/python/two_pointers/ThreeSumClosest_16.py
class ThreeSumClosest_16(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        closest = sum(nums[0:3])
        if len(nums) == 3:
            return closest
        nums.sort()
        for i in range(len(nums) - 2):
            start, end = i + 1, len(nums) - 1
            while start < end:
                tsum = nums[i] + nums[start] + nums[end]
                if tsum == target:
                    return tsum
                elif tsum < target:
                    start += 1
                elif tsum > target:
                    end -= 1
                if abs(tsum - target) < abs(closest - target):
                    closest = tsum
        return closest
